### PR TITLE
Deeper Backbone: 4 TransolverBlocks for increased model capacity

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

We are currently using only **46 GB of 96 GB available VRAM** — the model is significantly under-capacity. The stochastic depth experiment (PR #2230) demonstrated that all 3 TransolverBlocks are **load-bearing** — each handles a distinct aspect of the physics pipeline (geometry encoding, physics interaction, global pressure). This means the model needs MORE capacity, not less.

Adding a **4th TransolverBlock** gives the model:
1. **Redundancy** — blocks can specialize and develop deeper feature hierarchies
2. **More representational capacity** — more parameters to capture complex flow physics
3. **Better gradient flow** — 4 blocks with skip connections provide richer gradient paths than 3

In vision transformers, going from 3 to 4+ blocks is often the threshold where multi-scale feature learning emerges. The additional block allows the model to develop an intermediate representation between low-level geometry and high-level physics.

**VRAM estimate:** Each TransolverBlock adds ~25% to the backbone cost. Expected VRAM: ~55-60 GB (well within 96 GB limit).

**Epoch count adjustment:** With ~33% more compute per forward pass, the wall-clock timeout will reduce epochs from ~146 to ~110. The cosine schedule must be adjusted: `cosine_T_max=120` to complete the annealing cycle within the reduced epoch budget.

**Confidence:** Medium-High. This is a fundamental scaling experiment that hasn't been tried with the current feature stack. The VRAM headroom strongly suggests we're leaving capacity on the table.

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: No code changes needed

The `--n_layers` flag already exists and controls the number of TransolverBlocks. Simply change from 3 to 4 in the run command. The model architecture, skip connections, pressure_deep branches, and all other mechanisms should automatically adapt to the extra block.

**IMPORTANT verification:** Before running, verify that `--n_layers 4` works by checking that:
- The model instantiates without errors
- `pressure_deep` and `pressure_first` still work (they depend on block indices)
- The SRF heads still receive the correct backbone output
- If pressure_deep uses hardcoded block indices (e.g., `blocks[1]`), you may need to adjust

If pressure_deep or any other feature hardcodes block indices for `n_layers=3`, fix it to be `n_layers`-agnostic (e.g., use `blocks[n_layers//2]` for the intermediate supervision point).

### Step 2: Run 2 seeds with adjusted cosine schedule

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/deeper-backbone-s42" \
  --wandb_group "round19/deeper-backbone" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 4 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 120 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73 — identical but --seed 73 --wandb_name "alphonse/deeper-backbone-s73"
```

Key changes from baseline:
- `--n_layers 4` (was 3)
- `--cosine_T_max 120` (was 160, adjusted for fewer expected epochs)

### Step 3: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, comparison to baseline, W&B run IDs.

Pay special attention to:
- **Epoch count:** How many epochs completed before timeout? If <100, the model is undertrained.
- **VRAM usage:** Report peak VRAM. If >85 GB, there may be OOM risk on some configs.
- **Training curve shape:** Does the loss curve show the model is still improving at timeout? If so, the deeper model may need more training time (favorable sign).
- **Per-metric analysis:** The deeper model should help ALL metrics if capacity was the bottleneck. If only some metrics improve, the extra block may be learning domain-specific representations.

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: `hgml7i2r` (seed 42), `qic03vrg` (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Summary

4 TransolverBlocks regressed slightly vs baseline on 2-seed average, but with **very high seed variance** — seed 42 nearly matched baseline on p_re and p_tan, while seed 73 regressed substantially. The 4-layer model appears **undertrained** due to the slower per-epoch time (91s vs 73s for 3-layer) producing only 118-119 epochs vs ~145 for the 3-layer baseline.

### Metrics Table

| Metric | Baseline (3L) | Seed 42 (4L) | Seed 73 (4L) | 2-seed avg | vs Baseline |
|--------|---------------|-------------|-------------|------------|-------------|
| **p_in**   | 11.979 | 12.621 | 14.096 | **13.358** | +1.379 (+11.5%, worse) |
| **p_oodc** | 7.643  | 8.177  | 8.879  | **8.528**  | +0.885 (+11.6%, worse) |
| **p_tan**  | 28.341 | 28.884 | 28.624 | **28.754** | +0.413 (+1.5%, worse) |
| **p_re**   | 6.300  | 6.323  | 7.186  | **6.754**  | +0.454 (+7.2%, worse) |

**Note on seed variance:** The gap between seed 42 and seed 73 is very large (1.5 units on p_in). This suggests high instability in the 4-layer model under the wall-clock budget, not a clean negative result.

### Run Details

| | Seed 42 | Seed 73 |
|--|---------|---------|
| W&B run | `2xd7vf84` | `dqupxvd5` |
| Best epoch | 118 / 500 (wall-clock timeout) | 119 / 500 (wall-clock timeout) |
| VRAM | 54.8 GB | 54.9 GB |
| Epoch time | ~91s | ~89s |
| cosine_T_max | 120 | 120 |

For comparison, the 3-layer baseline: best epoch ~143-146, ~73s/epoch, 46.3 GB VRAM.

### Training Dynamics Observation

At epoch 112 (pre-timeout), seed 42 had val_in_dist=0.2840 and seed 73 had val_in_dist=0.2804 — both still improving. The 3-layer baseline's best val_in_dist was 0.2505-0.2628. The 4-layer model was **still on a declining trajectory** at timeout, suggesting it would continue to improve with more training.

### Command

```bash
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/deeper-backbone-s42" \
  --wandb_group "round19/deeper-backbone" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 4 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 120 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
# Seed 73 identical with --seed 73 --wandb_name "alphonse/deeper-backbone-s73"
```

### What Happened

1. **Undertrained, not conclusive:** The 4-layer model got ~118 epochs vs ~145 for the 3-layer baseline. At epoch 112, both seeds still had declining loss. With the wall-clock constraint, we're comparing an undertrained 4-layer model to a better-trained 3-layer model. This is an apples-to-oranges comparison.

2. **Seed 42 was nearly competitive on p_re and p_tan:** p_re=6.32 (baseline 6.30, δ=+0.02) and p_tan=28.88 (baseline 28.34, δ=+0.54). The 4-layer model may genuinely have the capacity to match or beat baseline if given sufficient training time.

3. **VRAM confirms capacity headroom:** 54.8 GB used (vs 46.3 GB for 3-layer). Still well within 96 GB. The model could be made even larger.

4. **High seed variance is a concern:** A 1.5 unit difference between seed 42 and seed 73 on p_in is unusually large. This may indicate the 4-layer model's loss landscape is less smooth/stable than the 3-layer version, requiring more careful LR tuning or a longer warmup.

### Suggested Follow-ups

- **Extended training time:** If SENPAI_TIMEOUT_MINUTES can be increased to 240+ min for this experiment, the 4-layer model would have ~160+ epochs and a fairer comparison with the 3-layer baseline. Without this, the capacity hypothesis can't be properly tested.
- **Adjusted cosine schedule:** `cosine_T_max=180` would be more appropriate for ~120 actual epochs (no cycle should complete at exactly the timeout). The current `cosine_T_max=120` likely causes premature LR collapse right at the end.
- **Smaller slice_num with 4 layers**: `--slice_num 64` would reduce memory per block and may allow ~130+ epochs at the same timeout. This trades some attention granularity for more training steps.
- **Better VRAM usage:** The model currently uses 54.8 GB / 96 GB. Could scale up `n_hidden` or `slice_num` along with `n_layers=4` to better utilize available capacity.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2240 because ADVISOR: Closing — all metrics regressed (p_in +11.5%, p_oodc +11.6%, p_tan +1.5%, p_re +7.2%). Your analysis about undertraining is correct: only 118 epochs vs ~145 for 3-layer, and still improving. However, the wall-clock budget is a hard constraint (SENPAI_TIMEOUT_MINUTES), so the 4-layer model simply can't get enough training within our allotted time. The high seed variance (+1.5 p_in between seeds) also suggests the deeper model's loss landscape is less stable. The 3-layer architecture is well-matched to our training budget. Thank you for the thorough analysis — the VRAM data (54.8GB) and still-declining loss curves are useful for future reference if the training budget changes..
